compiler/rustc_pattern_analysis/src/rustc.rs

```diff
@@ -226,6 +226,11 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
                         // The only legal patterns of type `Box` (outside `std`) are `_` and box
                         // patterns. If we're here we can assume this is a box pattern.
                         reveal_and_alloc(cx, once(args.type_at(0)))
+                    } else if adt.is_pin()
+                        && let ty::Ref(_, rty, _) = args.type_at(0).kind()
+                        && self.tcx.features().pin_ergonomics()
+                    {
+                        reveal_and_alloc(cx, once(*rty))
```

Comment:
That doesn't look like the correct logic. This code here would prevent e.g. `&42` working on `Pin<&u32>`, or anything working on `Pin<Box<T>>` even with deref patterns. Does the deref_patterns logic not handle this well already?

Reply:
Note that the thing above with `Box` is an old hack we should remove now that deref_patterns works.